src/properties.ts

```diff
@@ -0,0 +1,319 @@
+  }
+  return packageJsonParams;
+}
+
+/**
+ * Convert CLI args into scanner properties.
+ */
+function getCommandLineProperties(cliArgs?: string[]): ScannerProperties {
+  if (!cliArgs || cliArgs.length === 0) {
+    return {};
+  }
+
+  // Parse CLI args (eg: -Dsonar.token=xxx)
+  const properties: ScannerProperties = {};
+  for (const arg of cliArgs) {
+    if (!arg.startsWith('-D')) {
+      continue;
+    }
+    const [key, value] = arg.substring(2).split('=');
+    properties[key] = value;
+  }
+
+  return properties;
+}
+
+/**
+ * Parse properties stored in sonar project properties file, if it exists.
+ */
+function getSonarFileProperties(projectBaseDir: string): ScannerProperties {
+  // Read sonar project properties file in project base dir
+  try {
+    const sonarPropertiesFile = path.join(projectBaseDir, SONAR_PROJECT_FILENAME);
+    const properties: ScannerProperties = {};
+    const data = fs.readFileSync(sonarPropertiesFile).toString();
+    const lines = data.split('\n');
+    for (const line of lines) {
+      const trimmedLine = line.trim();
+      if (trimmedLine.length === 0 || trimmedLine.startsWith('#')) {
+        continue;
+      }
+      const [key, value] = trimmedLine.split('=');
+      properties[key] = value;
+    }
+
+    return properties;
+  } catch (error: any) {
+    log(LogLevel.WARN, `Failed to read ${SONAR_PROJECT_FILENAME} file: ${error.message}`);
+    throw error;
+  }
+}
+
+/**
+ * Get scanner properties from scan option object (JS API).
+ */
+function getScanOptionsProperties(scanOptions: ScanOptions): ScannerProperties {
+  const options = {
+    ...scanOptions.options,
+  };
+
+  if (typeof scanOptions.serverUrl !== 'undefined') {
+    options[ScannerProperty.SonarHostUrl] = scanOptions.serverUrl;
+  }
+
+  if (typeof scanOptions.token !== 'undefined') {
+    options[ScannerProperty.SonarToken] = scanOptions.token;
+  }
+
+  if (typeof scanOptions.verbose !== 'undefined') {
+    options[ScannerProperty.SonarVerbose] = scanOptions.verbose ? 'true' : 'false';
+  }
+
+  return options;
+}
+
+/**
+ * Automatically parse properties from environment variables.
+ */
+export function getEnvironmentProperties() {
+  const { env } = process;
+
+  const jsonEnvVariables = ['SONAR_SCANNER_JSON_PARAMS', 'SONARQUBE_SCANNER_PARAMS'];
+
+  let properties: ScannerProperties = {};
+
+  // Get known environment variables
+  for (const [envName, scannerProperty] of ENV_TO_PROPERTY_NAME) {
+    if (envName in env) {
+      properties[scannerProperty] = env[envName] as string;
+    }
+  }
+
+  // Get generic environment variables
+  properties = {
+    ...properties,
+    ...Object.fromEntries(
+      Object.entries(env)
+        .filter(([key]) => key.startsWith(ENV_VAR_PREFIX))
+        .filter(([key]) => !jsonEnvVariables.includes(key))
+        .map(([key, value]) => [envNameToSonarPropertyNameMapper(key), value as string]),
+    ),
```

Comment:
Yee although `value !== undefined` will never be false because `process.env` only returns string values, so to me the cast is what we should use here because it's just TS who is too dumb to get it, there's no safety issue? That's the type for `env`:
![image](https://github.com/SonarSource/sonar-scanner-npm/assets/31401273/cfd0ccec-be8e-4bd6-8f7f-a11961e0c0d1)

wdyt?

Reply:
you're right it won't be `undefined`, but it could be `'undefined'` 😅. 
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/process.d.ts#L534

I also think that it CAN be `undefined` but this ProcessEnv is casting it to string

```
lucas.paulger@MAC0343 sonar-scanner-npm % node -p "console.log(typeof process.env.NODE_ENV);"
undefined
```

I was expecting string 🤔

Reply:
I tested locally, I think it's overkill what I asked because the Object.entries won't pass keys with no assigned value. Marking as resolved. you can keep this as is